Testing.Rmd

```diff
@@ -596,5 +1364,101 @@ There are a few things to bear in mind:
+    checked on.
+    Depends on how broad the user base is and the "trickiness" of the package's
+    code (e.g. does it need compilation?).
+    Do a cost benefit analysis.
+    Connect that to which `usethis::use_github_action_check_*()` variant is
+    appropriate.
+* How to decide whether to let a test run on CRAN.
+  If the test can fail WITHOUT indicating a defect in your package that merits
+  removal from CRAN, do not let CRAN run the test. Examples:
+  - Accessing a website or API that could ever be down or have an
+    occasional error (so: any website or API on planet earth).
+  - Exact formatting of a message.
+  - File system work typical of an authentic user that touches files any where
+    but the R session's temp directory. This includes reading or writing the
+    clipboard.
+* How to prevent a test from running on CRAN.
+  - `skip_on_cran()`, `skip_on_os()`
+  - Default behaviour for snapshot tests
+  - Keep in a separate repo
+
+## Miscellaneous sections to (maybe) add
+
+### Mocking
+
+At the very least, I think we need to point to packages that handle this and outline situations when mocking is one of the reasonable approaches.
+
+It's not actually a topic I want to get deeply into, because it feels pretty specialized and I don't think we have detailed, strongly held, stable opinions about it.
+
+### Filesystem matters
+
+Do not write into `tests/testthat/` during your tests.
+
+The only place you should write into is session temp dir.
+
+Even then, best practice is to clean up any files or folder you create.
+
+Ways to be build good paths:
+
+* `testthat::test_path()`
+* `fs::path_package()` and `system.file()`
+
+### Namespace
+
+You can call / test all of your own package's functions without any special effort on your part, regardless of whether they are exported or not.
+
+That is provided at test time by `testhat::test_file()`, `devtools::test()`, etc.
+
+During interactive development, you should make this happen for yourself via `devtools::load_all()`.
+
+In general, testthat's official workflows are designed around the assumption that the user uses `devtools::load_all()`.
+
+### Dependencies
+
+Don't attach other packages from your tests.
+Don't call `library()`.
+Or, rather, that should be about as rare as using `Depends`.
+
+If other package is in Imports, do `foo::fun()`, just like we recommend when you use it in functions below `R/`.
+
+If other package is in Suggests, .... I already wrote about whether and how to guard in Metadata.Rmd.
+Move that content?
+
+### Introspection
+
+`NOT_CRAN`, `skip_on_cran()`
+
+`TESTTHAT`, `is_testing()`
+
+`testing_package()`
+
+### Test coverage
+
+* What is test coverage?
+* Use covr to learn your test coverage.
+* Use GitHub Actions to keep tabs on your test coverage.
+* Test coverage should be "high", but going from 90% to 100% coverage is not
+  always the best use of your development time and energy.
```

Comment:
Especially since the last 1% might require excessive gymnastics to cover. Don't sacrifice maintainability for covering some weird edge case that isn't that important.